lib/cc/analyzer/engine.rb

```diff
@@ -35,35 +35,34 @@ def run(stdout_io, stderr_io = StringIO.new)
           end
         end
 
-        Thread.new do
+        t_timeout = Thread.new do
           sleep TIMEOUT
-
-          if engine_running
-            Thread.current.abort_on_exception = true
-            run_command("docker kill #{container_name}")
-
-            Analyzer.statsd.increment("cli.engines.result.error")
-            Analyzer.statsd.increment("cli.engines.result.error.timeout")
-            Analyzer.statsd.increment("cli.engines.names.#{name}.result.error")
-            Analyzer.statsd.increment("cli.engines.names.#{name}.result.error.timeout")
-            raise EngineTimeout, "engine #{name} ran past #{TIMEOUT} seconds and was killed"
-          end
+          run_command("docker kill #{container_name} || true")
+          timed_out = true
         end
 
         pid, status = Process.waitpid2(pid)
-        engine_running = false
+        t_timeout.kill
+
         Analyzer.statsd.increment("cli.engines.finished")
 
-        if status.success?
+        if timed_out
+          Analyzer.statsd.increment("cli.engines.result.error")
+          Analyzer.statsd.increment("cli.engines.result.error.timeout")
+          Analyzer.statsd.increment("cli.engines.names.#{name}.result.error")
+          Analyzer.statsd.increment("cli.engines.names.#{name}.result.error.timeout")
+          raise EngineTimeout, "engine #{name} ran past #{TIMEOUT} seconds and was killed"
+        elsif status.success?
           Analyzer.statsd.increment("cli.engines.names.#{name}.result.success")
           Analyzer.statsd.increment("cli.engines.result.success")
         else
           Analyzer.statsd.increment("cli.engines.names.#{name}.result.error")
           Analyzer.statsd.increment("cli.engines.result.error")
-
           raise EngineFailure, "engine #{name} failed with status #{status.exitstatus} and stderr #{stderr_io.string.inspect}"
         end
       ensure
+        t_timeout.kill if t_timeout
+
```

Comment:
I actually put this here intentionally because I wanted to separate this from the other threads (which are handled differently, and are grouped together)